File: ios/HarnessMobile/UI/StreamingMarkdownBlocks.swift
import Foundation

/// Splits streaming markdown into stable completed blocks plus a trailing partial.
/// `completed.joined() + trailing == content` always holds.
struct StreamingMarkdownBlocks: Equatable {
    var completed: [String]
    var trailing: String

    static func split(_ content: String) -> StreamingMarkdownBlocks {
        split(content, previous: nil)
    }

    /// Incremental split: when `content` still begins with `previous.completed.joined()`,
    /// only re-scan the remainder so settled blocks stay identity-stable for MarkdownUI.
    static func split(_ content: String, previous: StreamingMarkdownBlocks?) -> StreamingMarkdownBlocks {
        if let previous, !previous.completed.isEmpty {
            let prefix = previous.completed.joined()
            if content.hasPrefix(prefix) {
                var remainder = String(content.dropFirst(prefix.count))
                var completed = previous.completed

                // Full split attaches blank-line separators to the preceding segment when a
                // new segment starts. Mirror that when growing past a fully-consumed trailing.
                if previous.trailing.isEmpty,
                   let absorbed = absorbLeadingBlankLineSeparators(from: &remainder),
                   !absorbed.isEmpty {
                    completed[completed.count - 1] += absorbed
                }

                let remainderSplit = splitFull(remainder)
                return StreamingMarkdownBlocks(
                    completed: completed + remainderSplit.completed,
                    trailing: remainderSplit.trailing
                )
            }
        }
        return splitFull(content)
    }

    /// If `text` starts with one or more blank lines and then more content, move those
    /// separator lines out of `text` and return them. EOF-only blank lines are left alone.
    private static func absorbLeadingBlankLineSeparators(from text: inout String) -> String? {
        guard !text.isEmpty else { return nil }
        var index = text.startIndex
        var absorbed = ""
        while index < text.endIndex {
            let lineEnd = text[index...].firstIndex(of: "\n").map { text.index(after: $0) } ?? text.endIndex
            let line = String(text[index..<lineEnd])
            let body = line.hasSuffix("\n") ? String(line.dropLast()) : line
            guard isBlankLine(body) else { break }
            absorbed += line
            index = lineEnd
        }
        guard !absorbed.isEmpty, index < text.endIndex else { return nil }
        text = String(text[index...])
        return absorbed
    }

    private static func splitFull(_ content: String) -> StreamingMarkdownBlocks {
        guard !content.isEmpty else {
            return StreamingMarkdownBlocks(completed: [], trailing: "")
        }

        let segments = fenceAwareSegments(in: content)
        guard let last = segments.last else {
            return StreamingMarkdownBlocks(completed: [], trailing: "")
        }

        if segments.count == 1 {
            if isClosedFenceBlock(last) {
                return StreamingMarkdownBlocks(completed: [last], trailing: "")
            }
            return StreamingMarkdownBlocks(completed: [], trailing: last)
        }

        var completed = Array(segments.dropLast())
        var trailing = last

        while let block = completed.last,
              !isClosedFenceBlock(block),
              hasUnbalancedInlineMarkers(block) {
            trailing = block + trailing
            completed.removeLast()
        }

        if isClosedFenceBlock(trailing) {
            completed.append(trailing)
            trailing = ""
        }

        return StreamingMarkdownBlocks(completed: completed, trailing: trailing)
    }

    /// Segments separated by blank lines outside fences. Each separator (`\n\n+`) stays
    /// attached to the preceding segment so joining reconstructs `content`.
    private static func fenceAwareSegments(in content: String) -> [String] {
        var segments: [String] = []
        var current = ""
        var inFence = false
        var lineStart = content.startIndex

        while lineStart < content.endIndex {
            let lineEnd = content[lineStart...].firstIndex(of: "\n").map { content.index(after: $0) }
                ?? content.endIndex
            let line = String(content[lineStart..<lineEnd])
            let lineBody = line.hasSuffix("\n") ? String(line.dropLast()) : line

            if isFenceDelimiter(lineBody) {
                inFence.toggle()
                current += line
                lineStart = lineEnd
                continue
            }

            if !inFence, isBlankLine(lineBody), !current.isEmpty {
                current += line
                lineStart = lineEnd
                while lineStart < content.endIndex {
                    let nextEnd = content[lineStart...].firstIndex(of: "\n").map { content.index(after: $0) }
                        ?? content.endIndex
                    let nextLine = String(content[lineStart..<nextEnd])
                    let nextBody = nextLine.hasSuffix("\n") ? String(nextLine.dropLast()) : nextLine
                    if isBlankLine(nextBody) {
                        current += nextLine
                        lineStart = nextEnd
                        continue
                    }
                    break
                }
                if lineStart < content.endIndex {
                    segments.append(current)
                    current = ""
                }
                continue
            }

            current += line
            lineStart = lineEnd
        }

        segments.append(current)
        return segments
    }

    private static func isFenceDelimiter(_ line: String) -> Bool {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard trimmed.hasPrefix("```") else { return false }
        let afterTicks = trimmed.drop(while: { $0 == "`" })
        return !afterTicks.contains("`")
    }

    private static func isBlankLine(_ line: String) -> Bool {
        line.trimmingCharacters(in: .whitespaces).isEmpty
    }

    static func isClosedFenceBlock(_ block: String) -> Bool {
        let lines = block.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        guard let firstNonBlank = lines.first(where: { !isBlankLine($0) }) else {
            return false
        }
        guard isFenceDelimiter(firstNonBlank) else { return false }

        var fenceCount = 0
        for line in lines {
            if isFenceDelimiter(line) {
                fenceCount += 1
            }
        }
        return fenceCount >= 2 && fenceCount % 2 == 0
    }

    /// Unbalanced `**` or inline `` ` `` (ignoring fence delimiter lines) → unsafe for MarkdownUI yet.
    static func hasUnbalancedInlineMarkers(_ text: String) -> Bool {
        if unbalancedMarkerCount(of: "**", in: text) { return true }
        if hasUnbalancedInlineBackticks(text) { return true }
        return false
    }

    private static func unbalancedMarkerCount(of marker: String, in text: String) -> Bool {
        var count = 0
        var search = text.startIndex
        while let range = text.range(of: marker, range: search..<text.endIndex) {
            count += 1
            search = range.upperBound
        }
        return count % 2 != 0
    }

    private static func hasUnbalancedInlineBackticks(_ text: String) -> Bool {
        var inlineTicks = 0
        for line in text.split(separator: "\n", omittingEmptySubsequences: false) {
            let body = String(line)
            if isFenceDelimiter(body) { continue }
            for ch in body where ch == "`" {
                inlineTicks += 1
            }
        }
        return inlineTicks % 2 != 0
    }
}

File: src/shared/streamingMarkdownBlocks.ts
/**
 * Splits streaming markdown into stable completed blocks plus a trailing partial.
 * completed.join("") + trailing === content always holds.
 */

export interface StreamingMarkdownBlocks {
  completed: string[];
  trailing: string;
}

export function splitStreamingMarkdown(
  content: string,
  previous?: StreamingMarkdownBlocks | null
): StreamingMarkdownBlocks {
  if (previous && previous.completed.length > 0) {
    const prefix = previous.completed.join("");
    if (content.startsWith(prefix)) {
      let remainder = content.slice(prefix.length);
      const completed = [...previous.completed];

      if (previous.trailing === "" && remainder.length > 0) {
        const absorbed = absorbLeadingBlankLineSeparators(remainder);
        if (absorbed) {
          completed[completed.length - 1] += absorbed.text;
          remainder = absorbed.rest;
        }
      }

      const remainderSplit = splitStreamingMarkdownFull(remainder);
      return {
        completed: completed.concat(remainderSplit.completed),
        trailing: remainderSplit.trailing,
      };
    }
  }
  return splitStreamingMarkdownFull(content);
}

/** Include trailing partial as a final block (stream end). */
export function flushStreamingMarkdown(blocks: StreamingMarkdownBlocks): StreamingMarkdownBlocks {
  if (!blocks.trailing) return blocks;
  return {
    completed: [...blocks.completed, blocks.trailing],
    trailing: "",
  };
}

/** True when the UI should show at least one completed markdown block. */
export function hasVisibleStreamingBlocks(content: string, isStreaming: boolean): boolean {
  const split = splitStreamingMarkdown(content);
  const presented = isStreaming ? split : flushStreamingMarkdown(split);
  return presented.completed.length > 0;
}

function absorbLeadingBlankLineSeparators(text: string): { text: string; rest: string } | null {
  if (!text) return null;
  let index = 0;
  let absorbed = "";
  while (index < text.length) {
    const nl = text.indexOf("\n", index);
    const lineEnd = nl === -1 ? text.length : nl + 1;
    const line = text.slice(index, lineEnd);
    const body = line.endsWith("\n") ? line.slice(0, -1) : line;
    if (!isBlankLine(body)) break;
    absorbed += line;
    index = lineEnd;
  }
  if (!absorbed || index >= text.length) return null;
  return { text: absorbed, rest: text.slice(index) };
}

function splitStreamingMarkdownFull(content: string): StreamingMarkdownBlocks {
  if (!content) {
    return { completed: [], trailing: "" };
  }

  const segments = fenceAwareSegments(content);
  const last = segments[segments.length - 1];
  if (!last) {
    return { completed: [], trailing: "" };
  }

  if (segments.length === 1) {
    if (isClosedFenceBlock(last)) {
      return { completed: [last], trailing: "" };
    }
    return { completed: [], trailing: last };
  }

  const completed = segments.slice(0, -1);
  let trailing = last;

  while (completed.length > 0) {
    const block = completed[completed.length - 1]!;
    if (isClosedFenceBlock(block) || !hasUnbalancedInlineMarkers(block)) break;
    trailing = block + trailing;
    completed.pop();
  }

  if (isClosedFenceBlock(trailing)) {
    completed.push(trailing);
    trailing = "";
  }

  return { completed, trailing };
}

function fenceAwareSegments(content: string): string[] {
  const segments: string[] = [];
  let current = "";
  let inFence = false;
  let lineStart = 0;

  while (lineStart < content.length) {
    const nl = content.indexOf("\n", lineStart);
    const lineEnd = nl === -1 ? content.length : nl + 1;
    const line = content.slice(lineStart, lineEnd);
    const lineBody = line.endsWith("\n") ? line.slice(0, -1) : line;

    if (isFenceDelimiter(lineBody)) {
      inFence = !inFence;
      current += line;
      lineStart = lineEnd;
      continue;
    }

    if (!inFence && isBlankLine(lineBody) && current.length > 0) {
      current += line;
      lineStart = lineEnd;
      while (lineStart < content.length) {
        const nextNl = content.indexOf("\n", lineStart);
        const nextEnd = nextNl === -1 ? content.length : nextNl + 1;
        const nextLine = content.slice(lineStart, nextEnd);
        const nextBody = nextLine.endsWith("\n") ? nextLine.slice(0, -1) : nextLine;
        if (isBlankLine(nextBody)) {
          current += nextLine;
          lineStart = nextEnd;
          continue;
        }
        break;
      }
      if (lineStart < content.length) {
        segments.push(current);
        current = "";
      }
      continue;
    }

    current += line;
    lineStart = lineEnd;
  }

  segments.push(current);
  return segments;
}

function isFenceDelimiter(line: string): boolean {
  const trimmed = line.trim();
  if (!trimmed.startsWith("```")) return false;
  const afterTicks = trimmed.replace(/^`+/, "");
  return !afterTicks.includes("`");
}

function isBlankLine(line: string): boolean {
  return line.trim().length === 0;
}

export function isClosedFenceBlock(block: string): boolean {
  const lines = block.split("\n");
  const firstNonBlank = lines.find((l) => !isBlankLine(l));
  if (!firstNonBlank || !isFenceDelimiter(firstNonBlank)) return false;

  let fenceCount = 0;
  for (const line of lines) {
    if (isFenceDelimiter(line)) fenceCount += 1;
  }
  return fenceCount >= 2 && fenceCount % 2 === 0;
}

export function hasUnbalancedInlineMarkers(text: string): boolean {
  if (unbalancedMarkerCount("**", text)) return true;
  if (hasUnbalancedInlineBackticks(text)) return true;
  return false;
}

function unbalancedMarkerCount(marker: string, text: string): boolean {
  let count = 0;
  let index = 0;
  while (index < text.length) {
    const found = text.indexOf(marker, index);
    if (found === -1) break;
    count += 1;
    index = found + marker.length;
  }
  return count % 2 !== 0;
}

function hasUnbalancedInlineBackticks(text: string): boolean {
  let inlineTicks = 0;
  for (const line of text.split("\n")) {
    if (isFenceDelimiter(line)) continue;
    for (const ch of line) {
      if (ch === "`") inlineTicks += 1;
    }
  }
  return inlineTicks % 2 !== 0;
}
